### DISCUSSION TITLE
Synchronization primitives for access context.bot_data

### DISCUSSION THREAD

**Discussion by LisichkinVladimir** [python-telegram-bot/python-telegram-bot#3976] [Q&A] [answered]
Hi!
I'm writing my first bot, so sorry if anything goes wrong :)
I use context.bot_data to store data common to all users of my bot.
Do I need to use synchronization primitives (for example asyncio.Lock) to access my variable in context.bot_data['my_boy_data'] ?

**Comment by lemontree210** [comment, marked as answer]
Hi! No, as long as you run PTB application in a standard way, as shown in the Wiki and the examples

**Reply by LisichkinVladimir** [reply]
Thank you

**Reply by Bibo-Joshi** [reply]
If you're making elaborate use of [concurrency](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Concurrency), then there might be cases where it makes sense, but I agree that for most standard use cases there is no need.

**Comment by LisichkinVladimir** [comment]
A question has arisen:
There is a standard PTB bot that, while is working, one user writes alone in context.bot_data['my_global_boy_data']
User two, will be able to read data from context.bot_data['my_global_boy_data'] ?
User two will be able to change global data if the need arises ?

**Reply by lemontree210** [reply]
you can make your code change `bot_data` whenever you need. As long as your bot is not persistent, you definitely shouldn't need to worry. However, I would recommend using `.user_data` for storing/changing user-related data.
`bot_data.some_attr` will be the same for all users, `user_data.some_attr` can be different for all users.

**Reply by LisichkinVladimir** [reply]
I'll have to explain what I want :)
There is a bot - a school schedule, which is parsed data from the site.
For 100% of users, the schedule class contains the same data.
But the schedule page may be change and I need to reload the schedule data without mistakenly.
So I'm worried that several clients may change the shared data.

**Reply by lemontree210** [reply]
Well if you provide no functionality for users to change the data (e.g. by asking them "what is the current schedule" and parsing their answers) then there is no way for them to affect the schedule you load from the website.

**Reply by LisichkinVladimir** [reply]
Every time I just get data from the page with the schedule, and if the data hash has changed, then I overload the global variable - schedule.
I'm worried that several users may overload data at the same time.

**Reply by lemontree210** [reply]
I wouldn't let the users trigger this. Instead, I would create a JobQueue (you can read about it in the Wiki) that makes sure the schedule is up-to-date and stored in `bot_data`. And let the users only receive the value of that `bot_data`. That way you drastically minimize the number of API requests.
